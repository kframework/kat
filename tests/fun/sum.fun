// sum program

letrec sum = fun 0 -> 0
             |   n -> n + sum (n - 1)
in sum 10

// 55
