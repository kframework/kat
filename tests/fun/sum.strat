(^ regular | #normal) *
(^ regular | #normal) * ; abstract
(^ regular | #normal) * ; abstract ; #loop
(^ regular | #normal) * ; abstract ; #loop ; (^ regular | #normal) *
(^ regular | #normal) * ; abstract ; #loop ; (^ regular | #normal) * ; which-can? #branch
(^ regular | #normal) * ; abstract ; #loop ; (^ regular | #normal) * ; ^ caseFailure ; (^ regular | #normal) *
(^ regular | #normal) * ; abstract ; #loop ; (^ regular | #normal) * ; ^ caseFailure ; (^ regular | #normal) 15 ; ^ switchFocus ; (^ regular | #normal) *
setAnalysis .Rules ; setStates .States ; push ; compile-step 5
setAnalysis .Rules ; setStates .States ; push ; compile-step 6
exec
