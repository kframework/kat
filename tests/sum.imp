// IMP Exmaples
// ============

// In this file several tests are provided for KAT. The file `tests/runtests.sh`
// should be run with K commit `e14da0a` from Github.

// Looping
// -------

// This is the classic `sum` program, which just sums the numbers from 1 to 10.


int n , s ;

n = 10 ;
while (0 <= n) {
  n = n + -1 ;
  s = s + n ;
}


// This version of the `sum` proogram has additional statements after the main
// while loop.

// Here the Collatz loop is provided, which calculates how many steps it takes a
// given number to go to 1 using Collatz update.
