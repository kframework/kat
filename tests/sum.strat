step-with skip ; step ; bimc 500 (bexp? s <= 32)
step-with skip ; step ; bimc 41  (bexp? s <= 32)
step-with skip ; step ; bimc 40  (bexp? s <= 32)
exec-to-branch
compile
