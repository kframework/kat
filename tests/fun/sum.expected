
Strategy: orig
================================================================================
<FUN>
  <k>
    55
  </k>
  <env>
    .Map
  </env>
  <envs>
    .List
  </envs>
</FUN>

Strategy: exec
================================================================================
<kat-FUN>
  <s>
    #STUCK ( )
  </s>
  <kat>
    <analysis>
      .Analysis
    </analysis>
    <states>
      .States
    </states>
  </kat>
  <harness>
    <FUN>
      <k>
        55
      </k>
      <env>
        .Map
      </env>
      <envs>
        .List
      </envs>
    </FUN>
  </harness>
</kat-FUN>

Strategy: (^ regular | #normal | #branch) *
================================================================================
<kat-FUN>
  <s>
    #STUCK ( )
  </s>
  <kat>
    <analysis>
      .Analysis
    </analysis>
    <states>
      .States
    </states>
  </kat>
  <harness>
    <FUN>
      <k>
        mu ( .Names , n + sum ( n - 1 ) ) ~> popEnv ~> popEnv ~> popEnv
      </k>
      <env>
        n |-> 10
        sum |-> closure ( .Map , 0 -> mu ( sum , .Names , 0 ) | n -> mu ( sum , .Names , n + sum ( n - 1 ) ) | .Cases )
      </env>
      <envs>
        ListItem ( .Map )
        ListItem ( sum |-> closure ( .Map , 0 -> mu ( sum , .Names , 0 ) | n -> mu ( sum , .Names , n + sum ( n - 1 ) ) | .Cases ) )
        ListItem ( .Map )
      </envs>
    </FUN>
  </harness>
</kat-FUN>

Strategy: # (^ regular | #normal | #branch) * ; abstract
================================================================================
SKIPPING

Strategy: # (^ regular | #normal | #branch) * ; abstract ; #loop
================================================================================
SKIPPING

Strategy: # (^ regular | #normal | #branch) * ; abstract ; #loop ; (^ regular | #normal) *
================================================================================
SKIPPING

Strategy: # (^ regular | #normal | #branch) * ; abstract ; #loop ; (^ regular | #normal) * ; which-can? #branch
================================================================================
SKIPPING

Strategy: # (^ regular | #normal | #branch) * ; abstract ; #loop ; (^ regular | #normal) * ; #caseFailure ; (^ regular | #normal) *
================================================================================
SKIPPING

Strategy: # (^ regular | #normal | #branch) * ; abstract ; #loop ; (^ regular | #normal) * ; #caseFailure ; (^ regular | #normal) 15 ; ^ switchFocus ; (^ regular | #normal) *
================================================================================
SKIPPING

Strategy: setAnalysis .Rules ; setStates .States ; push ; compile-step 5
================================================================================
<kat-FUN>
  <s>
    #STUCK ( )
  </s>
  <kat>
    <analysis>
      .Rules
    </analysis>
    <states>
      <FUN>
        <k>
          letrec ( sum = fun 0 -> 0 | n -> n + sum ( n - 1 ) | .Cases ) and .Bindings in sum 10
        </k>
        <env>
          .Map
        </env>
        <envs>
          .List
        </envs>
      </FUN> < ( ^ regular | ^ lookup | ^ applicationFocusFunction | ^ applicationFocusArgument | ^ listAssignment | ^ assignment | ^ letBinds | ^ letRecBinds ) * > < ^ caseIntFailure > < ( ^ regular | ^ lookup | ^ applicationFocusFunction | ^ applicationFocusArgument | ^ listAssignment | ^ assignment | ^ letBinds | ^ letRecBinds ) * > < ^ caseNameSuccess > < ( ^ regular | ^ lookup | ^ applicationFocusFunction | ^ applicationFocusArgument | ^ listAssignment | ^ assignment | ^ letBinds | ^ letRecBinds ) * > : .States
    </states>
  </kat>
  <harness>
    <FUN>
      <k>
        pushEnv ~> setEnv ( .Map ) ~> let ( n = 10 ) and .Bindings in mu ( sum , .Names , n + sum ( n - 1 ) ) ~> popEnv ~> popEnv
      </k>
      <env>
        sum |-> closure ( .Map , 0 -> mu ( sum , .Names , 0 ) | n -> mu ( sum , .Names , n + sum ( n - 1 ) ) | .Cases )
      </env>
      <envs>
        ListItem ( .Map )
      </envs>
    </FUN>
  </harness>
</kat-FUN>

Strategy: # setAnalysis .Rules ; setStates .States ; push ; compile-step 6
================================================================================
SKIPPING
